docs/source/extensions_tutorial/3_spec_api.rst

```diff
@@ -0,0 +1,173 @@
+The Spec API
+------------
+
+The `NWB Specification Language <https://schema-language.readthedocs.io/en/latest/index.html>`_ defines a structure for
+data and metadata using Groups, Datasets, Attributes, and Links. These structures are mapped onto
+:py:class:`~pynwb.spec.NWBGroupSpec`, :py:class:`~pynwb.spec.NWBDatasetSpec`,
+:py:class:`~pynwb.spec.NWBAttributeSpec`, and :py:class:`~pynwb.spec.NWBLinkSpec`, respectively. Here, we describe in
+detail each of these classes, and demonstrate how to use them to create custom neurodata types.
+
+Group Specifications
+^^^^^^^^^^^^^^^^^^^^
+
+Most neurodata types are Groups, which act like a directory or folder within the NWB file. A Group can have
+within it Datasets, Attributes, Links, and/or other Groups. Groups are specified with the
+:py:class:`~pynwb.spec.NWBGroupSpec` class, which provides a python API for defining
+`NWB Group <https://schema-language.readthedocs.io/en/latest/specification_language_description.html#groups>`_ objects.
+
+.. code-block:: python
+
+    from pynwb.spec import NWBGroupSpec
+
+    spec = NWBGroupSpec(
+        neurodata_type_def='MyType',
+        neurodata_type_inc='NWBDataInterface',
+        doc='A custom NWB type',
+        name='quux',
+        attributes=[...],
+        datasets=[...],
+        groups=[...],
+        links=[...]
+    )
+
+``neurodata_type_def`` and ``neurodata_type_inc`` define the neurodata type with the following rules:
+
+- ``neurodata_type_def`` declares the name of the neurodata type.
+- ``neurodata_type_inc`` indicates what data type you are extending (Groups must extend Groups, and Datasets must extend Datasets).
+- To define a new neurodata type that does not extend an existing type, use ``neurodata_type_def`` and not ``neurodata_type_inc``.
+- To use a type that has already been defined, use ``neurodata_type_inc`` and not ``neurodata_type_def``.
+- You can define a group that is not a neurodata type by omitting both ``neurodata_type_def`` and ``neurodata_type_inc``.
+
+.. tip::
+    Although you have the option not to, there are several advantages to defining new groups and neurodata types.
+    Neurodata types can be reused in multiple places in the schema, and can be linked to, while groups that are not
+    neurodata types cannot. You can also have multiple neurodata type groups of the same type in the same group,
+    whereas groups that are not neurodata types are limited to 0 or 1. Most of the time, we would recommend making a
+    group a neurodata type. It is also generally better to extend your neurodata type from an existing type. Look
+    through the `NWB schema <https://nwb-schema.readthedocs.io/en/latest/>`_ to see if a core neurodata type would
+    work as a base for your new type. If no existing type works, consider extending
+    :py:class:`~pynwb.base.NWBDataInterface``, which allows you to add the object to a processing module.
+
```

Comment:
```suggestion
.. tip::
     New neurodata types should always be declared at the top level of the schema rather than nesting type
     definitions. I.e., when creating a new neurodata type it should be placed at the top level of your schema
     and then included at the appropriate location via ``neurodata_type_inc``. This approach greatly simplifies
     management of types. 
```